cosmwasm/contracts/ibc-translator/src/reply.rs

```diff
@@ -0,0 +1,197 @@
+
+pub fn handle_complete_transfer_reply(
+    deps: DepsMut,
+    env: Env,
+    msg: Reply,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    // we should only be replying on success
+    ensure!(
+        msg.result.is_ok(),
+        "msg result is not okay, we should never get here"
+    );
+
+    let res_data_raw = cw_utils::parse_reply_execute_data(msg)
+        .context("failed to parse protobuf reply response_data")?
+        .data
+        .context("no data in the response, we should never get here")?;
+    let res_data: CompleteTransferResponse =
+        from_binary(&res_data_raw).context("failed to deserialize response data")?;
+    let contract_addr = res_data
+        .contract
+        .context("no contract in response, we should never get here")?;
+
+    // load interim state
+    let transfer_info = CURRENT_TRANSFER
+        .load(deps.storage)
+        .context("failed to load current transfer from storage")?;
+
+    // delete interim state
+    CURRENT_TRANSFER.remove(deps.storage);
+
+    // deserialize payload into the type we expect
+    let payload: GatewayIbcTokenBridgePayload = serde_json_wasm::from_slice(&transfer_info.payload)
+        .context("failed to deserialize transfer payload")?;
+
+    match payload {
+        GatewayIbcTokenBridgePayload::Simple {
+            chain,
+            recipient,
+            fee: _,
+            nonce: _,
+        } => {
+            let recipient_decoded = String::from_utf8(recipient.to_vec())
+                .context(format!("failed to convert {recipient} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                recipient_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                None,
+            )
+        }
+        GatewayIbcTokenBridgePayload::ContractControlled {
+            chain,
+            contract,
+            payload,
+            nonce: _,
+        } => {
+            let contract_decoded = String::from_utf8(contract.to_vec())
+                .context(format!("failed to convert {contract} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                contract_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                Some(payload),
+            )
+        }
+    }
+}
+pub fn convert_cw20_to_bank_and_send(
+    deps: DepsMut,
+    env: Env,
+    recipient: String,
+    amount: u128,
+    contract_addr: String,
+    chain_id: u16,
+    payload: Option<Binary>,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    // check the recipient and contract addresses are valid
+    // recipient will have a different bech32 prefix and fail
+    //deps.api
+    //    .addr_validate(&recipient)
+    //    .context(format!("invalid recipient address {}", recipient))?;
```

Comment:
good callout - we can remove this code and just keep the comment for validating the cw20 contract address